From 16ceeff642495fd16d62a890eec7e0a43f8ab8ba Mon Sep 17 00:00:00 2001
From: Charlene Yan <cyan@chromium.org>
Date: Tue, 17 Aug 2021 15:53:41 +0000
Subject: [PATCH 09/28] Cap boundaries so tabs can't be moved across
 {un}/pinned.

(cherry picked from commit 05198700faf344500e765218ca96b161b010fc45)

Bug: 1210985
Change-Id: I211c42703ebb384d019e1c8cda3d055d034e8492
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2945337
Commit-Queue: Charlene Yan <cyan@chromium.org>
Cr-Original-Commit-Position: refs/heads/master@{#891409}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3059451
Commit-Queue: Achuith Bhandarkar <achuith@chromium.org>
Reviewed-by: Achuith Bhandarkar <achuith@chromium.org>
Reviewed-by: Jana Grill <janagrill@google.com>
Owners-Override: Achuith Bhandarkar <achuith@chromium.org>
Owners-Override: Jana Grill <janagrill@google.com>
Cr-Commit-Position: refs/branch-heads/4430@{#1566}
Cr-Branched-From: e5ce7dc4f7518237b3d9bb93cccca35d25216cbe-refs/heads/master@{#857950}
---
 chrome/browser/ui/tabs/tab_strip_model.cc        | 12 +++++++++---
 .../browser/ui/tabs/tab_strip_model_unittest.cc  | 16 ++++++++++++++++
 2 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/chrome/browser/ui/tabs/tab_strip_model.cc b/chrome/browser/ui/tabs/tab_strip_model.cc
index 4b66f349e..7a986b53a 100644
--- a/chrome/browser/ui/tabs/tab_strip_model.cc
+++ b/chrome/browser/ui/tabs/tab_strip_model.cc
@@ -1939,13 +1939,18 @@ void TabStripModel::SelectRelativeTab(bool next, UserGestureDetails detail) {
 
 void TabStripModel::MoveTabRelative(bool forward) {
   const int offset = forward ? 1 : -1;
-
-  // TODO: this needs to be updated for multi-selection.
   const int current_index = active_index();
   base::Optional<tab_groups::TabGroupId> current_group =
       GetTabGroupForTab(current_index);
 
-  int target_index = std::max(std::min(current_index + offset, count() - 1), 0);
+  const int first_non_pinned_tab_index = IndexOfFirstNonPinnedTab();
+  int first_valid_index =
+      IsTabPinned(current_index) ? 0 : first_non_pinned_tab_index;
+  int last_valid_index =
+      IsTabPinned(current_index) ? first_non_pinned_tab_index - 1 : count() - 1;
+  int target_index = std::max(
+      std::min(current_index + offset, last_valid_index), first_valid_index);
+
   base::Optional<tab_groups::TabGroupId> target_group =
       GetTabGroupForTab(target_index);
 
@@ -1970,6 +1975,7 @@ void TabStripModel::MoveTabRelative(bool forward) {
       }
     }
   }
+  // TODO: this needs to be updated for multi-selection.
   MoveWebContentsAt(current_index, target_index, true);
 }
 
diff --git a/chrome/browser/ui/tabs/tab_strip_model_unittest.cc b/chrome/browser/ui/tabs/tab_strip_model_unittest.cc
index 41e2fa53f..7843b0858 100644
--- a/chrome/browser/ui/tabs/tab_strip_model_unittest.cc
+++ b/chrome/browser/ui/tabs/tab_strip_model_unittest.cc
@@ -2562,6 +2562,22 @@ TEST_F(TabStripModelTest, MoveTabNext_Group) {
   strip.CloseAllTabs();
 }
 
+TEST_F(TabStripModelTest, MoveTabNext_PinnedDoesNotGroup) {
+  TestTabStripModelDelegate delegate;
+  TabStripModel strip(&delegate, profile());
+  ASSERT_NO_FATAL_FAILURE(PrepareTabstripForSelectionTest(&strip, 4, 1, "0"));
+  EXPECT_EQ("0p 1 2 3", GetTabStripStateString(strip));
+
+  strip.AddToNewGroup({1, 2});
+  EXPECT_EQ(strip.GetTabGroupForTab(0), base::nullopt);
+
+  strip.MoveTabNext();
+  EXPECT_EQ("0p 1 2 3", GetTabStripStateString(strip));
+  EXPECT_EQ(strip.GetTabGroupForTab(0), base::nullopt);
+
+  strip.CloseAllTabs();
+}
+
 TEST_F(TabStripModelTest, MoveTabPrevious) {
   TestTabStripModelDelegate delegate;
   TabStripModel strip(&delegate, profile());
-- 
2.20.1

